### ISSUE TITLE
[1.1.38] Failed to connect to remote collector.

### ISSUE THREAD

**Issue by jhonnyx82** [Cacti/cacti#1604]
Hi Everyone,

devices from one of my remote pollers started to give this output on the device status (in EDIT device) but the devices themselves looks to be OK...do you have any clue if this is a bug or to fix this somehow?

![image](https://user-images.githubusercontent.com/26431888/39632861-8f5264ce-4fb6-11e8-881b-6e9f6c039ba1.png)

![image](https://user-images.githubusercontent.com/26431888/39632868-9712c01e-4fb6-11e8-8df2-95c6d28934cb.png)

**Comment by netniV** [comment]
This means that your main poller is having issues trying to connect using the same protocol to the remote_agent.php script on your poller or that it is timing out after 4-5 seconds of doing so.

**Comment by jhonnyx82** [comment]
the timeout would be related to the main poller to the remote poller http server? is this through http? timeout may be the case but was something changed in the latest releases ? I did not have this problem before and network performance looks the same I always had

even if 4-5 seconds looks a little too much to me...I have top 220 msec of RTT between the two..
I tried restarting apache and mariadb but no result

**Comment by netniV** [comment]
It's a http connection to the remote sites remote_agent.php file.  You'll find the error being generated in functions.php which will give you an indication of the URL that is used to call the remote poller.  So, timeout would be RTT + script execution time.

**Comment by jhonnyx82** [comment]
Thanks, 
I would like to add I find no errors in the apache errors_log and the access_log has equal entries like other pollers that works fine.

Now about the execution time: I am using snmp uptime, so I should get an snmp sysdesc and even on a device hosted in the same LAN (so same site) of the remote poller, I have the same timeout :/

somewhere where should I look? (function.php, I didn't find in the selective file debugs)

**Comment by netniV** [comment]
I'll take a look later as to whether I can add an option to debug remote poller calls (on a poller basis).  Really that kind of functionality should go into 1.2 but it shouldn't be difficult to back-port.

**Comment by netniV** [comment]
I can create a patch that has some debug info, but I'm by no means any expert on remote pollers. I haven't ever really used them in anger yet.

**Comment by jhonnyx82** [comment]
hi,

Still I am experiencing this problem even if snmpwalk and snmpget works OK from the remote poller shell, do you know where should I look at to resolve this ?

**Comment by cigamit** [comment]
You need to place some instrumentation in the remote agent, or view your apache access_log for remote agent calls on the remote system.  If it's an snmpv3 device, let us know, and then uninstall php-snmp on both sides restarting apache afterwards.

**Comment by jhonnyx82** [comment]
Hi,

 

·         Sorry, what type of instrumentation?

·         Apache access_log shows the same information as my other 4 remote pollers

·         No snmpv3, at the moment snmpv2

·         “uninstall php-snmp on both sides restarting apache afterwards” : is this just in case of snmpv3? Will this impact polling for the devices? “both ends” means on the main poller as well? In case I have no problem with other pollers, should I do that?

**Comment by jhonnyx82** [comment]
Hi,

While waiting for an answer on this, I would like to add that it appears that this started since my upgrade from version 1.1.36 to 1.1.38

Before all poller worked well and I just noticed that this impact realtime graph as well, probably you may think that this is my personal issue but you may have a bug, I am not sure if anyone reported this but this looks to be strange enough if not.

Thanks

**Comment by cigamit** [comment]
I've just retested the user interface with 1.2 and everything is working.  Verify your hostnames on the Data Collectors pages.

**Comment by cigamit** [comment]
You might have an SE Linux thing going on too.  It's saying it can not connect.  It may mean that the hostname in the Data Collectors page is not reachable.  Make certain that apache is reachable from the collector attempting to query the remote system.  It could be apache is down, or you can not connect to 80 or 443 on the hostname provided in the Data Collector.

**Comment by jhonnyx82** [comment]
no SELinux and the system can be reached for opening HTTP requests, I also see the requests coming in on the apache logs.

funny is that data collection works so it should be related only to the remote agent

**Comment by cigamit** [comment]
Edit your main poller and make sure it's hostname is not localhost.  It will poll the device directly since it's the same poller id as the current data collector.  You can verify that by tailing the Apache access log.

**Comment by eschoeller** [comment]
I'm having the same problem. I can't verify if this was working prior to 1.1.38. One thing I have discovered from reading through this post is that I did not have HTTPS enabled on my remote collectors - so I've fixed that but still have the same problem. My configuration may be somewhat unique ... my remote collectors reside on an isolated unrouted network behind the main poller. They masquerade all their traffic through the main poller's primary IP address. This configuration is really for adding more polling horsepower to Cacti, not really leveraging the "remote" aspect. I have DNAT rules setup on the main poller so I can access each HTTP interface through separate port numbers (aka port 8002 redirects to the internal IP of "poller 1" on port 80) 
All the polling functions work just fine, I just can't refresh data queries or anything unless I temporarily re-assign the device to the main poller, make any changes, then re-assign it back to the "remote" poller

**Comment by eschoeller** [comment]
I also found that I had unqualified names entered into the "Web Site Hostname". I switched these to static IPs. Not sure how things were working, and hopefully this doesn't break something. Now I get this message:
FATAL: You are not authorized to use this service
And in a tcpdump I actually see the main poller trying to hit port 80 on the remote poller instead of port 443. 
And I see this in the access log:
"GET /cacti/remote_agent.php?action=ping&host_id=649 HTTP/1.0" 200 216 "-" "-"

**Comment by netniV** [comment]
The system will always attempt to revert to a hostname from what I noticed in the code.  This is primarily because most systems using HTTPS do not specify an IP address in the certificate being used but a domain name.  

I will see if I have time at the weekend to setup a test remote collection and describe how I did it. I need to get a better understanding of it anyway.

**Comment by netniV** [comment]
One thing you can do is make sure that if your DNS doesn't resolve the remote collectors name, put it in /etc/hosts

**Comment by cigamit** [comment]
As an added security feature, we take the reverse dns from the request IP address of the requester.  It has to be one of the data collector hostnames, if the dns name that comes back does not match what is stored in the remote data collector hostnames, we reject the request.

**Comment by cigamit** [comment]
Are these servers behind a nat?  That might be the issue.

**Comment by eschoeller** [comment]
Yes, all of my "remote" collectors actually NAT through the main poller. They have no other connectivity besides that. The network is physically separate and dedicated. I haven't been working on it in awhile. It's semi-stable. Odd things happen from time to time - and I do see this particular message come back once in awhile. I'm still running on 1.1.38 with a few cherry-picked patches, but I'm so far off the current code base I can't merge patches in very cleanly anymore. And it's harder to test development releases now that this infrastructure spans more than one machine (while still being able to keep production polling up). Anyway, I'm really looking forward to an official 1.2 release so I can get back to tinkering:)

**Comment by cigamit** [comment]
Having @netniV on board has really accelerated things.  After 1.2, we are going to need a break.  I've already burned out a couple of times in the last few years.

**Comment by cigamit** [comment]
I did some re-ordering on the NAT reverse verification.  See if that helps.  Easy to cherry-pick.

**Comment by jhonnyx82** [comment]
Hi Everyone, long time no see...

Well in my network the remote pollers are **not** behind NAT bud are separated by a firewall and I had a constant look at the packet in transit (ran several packet capture) and I can see that port 80 is queried and used (so no HTTPS).

The funny part is that everything is working (querying the device and refreshing the data sources) just the SNMP description on the top of the device page seems to not work, also the reported SNMP state is OK

**Comment by jhonnyx82** [comment]
Also this is a successful query from the main poller to the remote poller (from apache access_log)

`[10/Aug/2018:15:34:54 +0000] "GET /cacti//remote_agent.php?action=runquery&host_id=2546&data_query_id=1 HTTP/1.0" 200 20511 "-" "-"`

this one instead is the query for the device status that fails :  

`[10/Aug/2018:15:38:04 +0000] "GET /cacti/remote_agent.php?action=ping&host_id=2546 HTTP/1.0" 200 488 "-" "-"`

**Comment by eschoeller** [comment]
Yes, I agree everything seems to be working but the SNMP description on the device page (and it's now intermittent for me). And I do see this HTTP issue as quite the security risk for anyone not paying too close of attention. In my case I don't mind, but if I was actually using pollers off on other parts of our network I would be very concerned. I'll try to cherry-pick the patch soon and see what changes I have.

**Comment by eschoeller** [comment]
I have applied the patch. Things are still working ok:) I'll report back anything I notice.

**Comment by netniV** [comment]
The double slash may be being interpreted as removing the /cacti/ part of the URL string.  Is cacti in a subfolder or not?  If not, the configuration is wrong somewhere.

**Comment by jhonnyx82** [comment]
Default configuration, in /cacti/ otherwise the first remote poller call wouldn't work either, not ?

**Comment by cigamit** [comment]
Edit the main collector, ensure the hostname is the real hostname of the main collector, from the remote data collectors do a reverse lookup of that string, and make sure it matches the IP Address of the interface talking from the main data collector.  If not, create a /etc/hosts entry for that IP Address back to the hostname saved in the hostname field of the main data collector.

**Comment by cigamit** [comment]
We are now checking not the IP address of the NAT, but of the source IP from the other side of the NAT.

**Comment by jhonnyx82** [comment]
> Edit the main collector, ensure the hostname is the real hostname of the main collector, from the remote data collectors do a reverse lookup of that string, and make sure it matches the IP Address of the interface talking from the main data collector. If not, create a /etc/hosts entry for that IP Address back to the hostname saved in the hostname field of the main data collector.

Does it matter that other pollers not beyond the firewall (that are also **not** NATed) works properly with the current setup?

If no NAT is involved why should I have a problem with the reverse lookup if all my remote pollers are synchronized on /etc/hosts file?

**Comment by jhonnyx82** [comment]
Hi,

Do you need some additional data or something? I can confirm that still 2 pollers over 5 have this issue, thank God at least the boost process is working so I can read the graphs from the main poller at least

**Comment by netniV** [comment]
Hi @jhonnyx82, to be clear, are you running 1.1.38 or 1.2 development?

**Comment by jhonnyx82** [comment]
1.1.38

**Comment by netniV** [comment]
The remote address has been addressed in 1.2 not 1.1. So you wouldn’t see any difference for now.

**Comment by jhonnyx82** [comment]
Thank you,

Can you let me know if the update process from 1.1.38 to 1.2.x with remote pollers is best performed in which fashion?

- primary poller only?
- primary poller then remote pollers?
- remote pollers then primary poller?

**Comment by eschoeller** [comment]
That is a great question, I’ve been shutting everything down upgrading the main poller then doing each remote poller before bringing them back up

**Comment by jhonnyx82** [comment]
@eschoeller  if you are using the 1.2.x version and realtime graph, can you let me know if you have still problems with the SNMP device info in the device panel and with realtime graphing from remote pollers?

**Comment by eschoeller** [comment]
I am unfortunately not on 1.2.x I’m nervous to jump to it until an official release. I’m not sure how I could test drive it in my environment.

**Comment by jhonnyx82** [comment]
This calls up for an integrated backup function (!!!) in CACTI

I know that CACTI is seen among the "legacy" NMS but IMHO is still effective and has a great potential, technically you can even expand it beyond simple SNMP checks but I would like to see more old "power features" like Weathermap along with a larger repository of templates and scripts (and of course some functions for management like backup, etc).

Anyway Thanks to the developers!

**Comment by netniV** [comment]
My suggest? Create a test environment, with test remote pollers. Add problematic devices and replicate the issue within that test environment.  Once you have it replicated, migrate the environment from 1.1.38 to 1.2 and see what the differences are.

**Comment by thurban** [comment]
Hi, not sure if this may be related to this issue. Nevertheless, I rolled out about 33 remote poller around the world, and I noticed the following:
- Pollers with good response times do not have any issues
- Pollers with not-so-good reponse times do have issues

I checked on one poller which is a very long distance away and realized the following:
- Startup of anything is pretty slow, once started it's fast
- Calling the include/global.php file in online mode is slow (6 to 8 secs)
- The remote poller website is very slow as well ( 6 to 8 secs before anything shows up )

I played around with the code and realized the following:
- Reducing the number of Script servers speeds up the poller startup ( 10 Script servers -> 60 to 80 seconds before it goes into polling mode ) -> 1 Script Server -> 6 to 8 seconds before it goes into polling mode
- global.php can be speed up to 2 second by forcing the read_config_option in include/functions.php to use the $local_db_cnn_id

From what I can see, the issue is with the remote pollers connecting to the remote database (Master) instead of using the local(synced) data for generic things.

---> The remote interface (Failed to connect to the remote poller) is happening due to a timeout occuring as the specific function has a (fixed) 5 second timeout ( lib/api_device.php -> $fgc_contextoption = get_default_contextoption(5); )

**Comment by netniV** [comment]
That last part would indeed raise the question of the SNMP contacting information spinning and failing.  5 < 6 which would be an obvious fail.  

For the database side, like you have said, having the pollers connect to a local DB would always be far faster and we had hoped to have remote pollers using their own data in 1.2, but I believe that hasn't happened due to some technical issues and that we really want to get 1.2 out the door in order to get a release out (there's been none for about 6 months).  So that's been pushed back a little while longer to 1.3.

**Comment by jhonnyx82** [comment]
Hi,

I would like to let you know that I tried https://github.com/Cacti/cacti/commit/5f22e34020fe1d2b7f7f3a3437890af6717ad890 and after also the latest rempote_poller.php into my cacti 1.1.38 code but it didn't work for me.

I guess this is confirmed as resolved for the full version 1.2.x ? or should we have a look again at this one ?

**Comment by netniV** [comment]
You will need the full 1.2 code for anything that works on the pollers tables as they have had new fields added, etc.  I'm not sure that there's any glaringly obvious bugs that would prevent you from trying out 1.2 Beta 0 (develop code) with a test server and test poller.  Beta 1 (released beta) will hopefully be prepared soon.

**Comment by max6491** [comment]
I have 1 master and 8 pollers.,. I am running cacti-1.1.38. Main poller seeing this message when I click on the device which is associated with non master poller. "ERROR: Failed to connect to remote collector." 
if I reassign the device to Master poller .. I don't see that message.
I tried checking the logs but no clue. also checked the spine and cacti configs ..all  of them seems to be correct, need some advise .. how to fix it.
Thanks
Max

**Comment by netniV** [comment]
If this is in the top left corner, that is because the system will attempt to communicate with the remote poller to query the status of the device in question.  I'm not sure without looking later if there are other messages based on the SSL cert not being valid etc that would be displayed. @cigamit will probably know better than me.

**Comment by max6491** [comment]
I do not have SSL configured on Cacti or any other pollers.

**Comment by netniV** [comment]
Sorry, should be a bit clearer, there is an RSA key that gets generated to encrypt/sign communication between pollers.  It was that process that I wondered if reported failures outside of just "can't connect".  Unfortunately, that is an area that I am a little cloudy on still as I haven't played with remote pollers in anger yet.

**Comment by netniV** [comment]
I wouldn't say I shamed you, just whispered very loudly that we should get it in...and finally you finished what you already started. ;-)

**Comment by eschoeller** [comment]
I'm excited!!

**Comment by cigamit** [comment]
Okay, so I created a timeout option now.  Default is 5 seconds for everything.  1.2 of course.

**Comment by jhonnyx82** [comment]
thanks, will give a try since I have still code 1.1.38.

Can I ask you to snip me a sample of adding the timeout ?  
I assume that is only on the **remote_agent.php** , right ?

 Sorry to make you waste time for this, just I'd like to not break things before I move to 1.2.x

**Comment by cigamit** [comment]
`cd /var/www/html/cacti
grep -r "get_default_contextoption(" *`

edit each file and make the call consistent.  The number in the middle is the timeout.
